prow/interrupts/interrupts.go

```diff
@@ -0,0 +1,69 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package interrupts exposes helpers for graceful handling of interrupt signals
+package interrupts
+
+import (
+	"context"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// RunOrShutdown runs the HTTP server and handles shutting it down
+// gracefully on interrupts. This function is meant to be called in
+// a blocking manner.
+func RunOrShutdown(server *http.Server, gracePeriod time.Duration, logger *logrus.Logger) {
+	go Shutdown(server, gracePeriod, logger)
+	logger.WithError(server.ListenAndServe()).Warn("Server exited.")
+}
+
+// Shutdown will shut down the server when an interrupt is received.
+// This function is meant to be called in a non-blocking manner.
+func Shutdown(server *http.Server, gracePeriod time.Duration, logger *logrus.Logger) {
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, os.Interrupt, syscall.SIGTERM)
+
+	<-sig
+	logger.Info("Server shutting down...")
+	ctx, cancel := context.WithTimeout(context.Background(), gracePeriod)
+	defer cancel()
+	if err := server.Shutdown(ctx); err != nil {
+		logger.WithError(err).Warn("Error shutting down server...")
+	}
+}
+
+// Tick will do work on an interval until an interrupt is received.
+// This function is meant to be called in a non-blocking manner.
+func Tick(work func(), interval time.Duration, logger *logrus.Logger) {
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, os.Interrupt, syscall.SIGTERM)
+
+	for {
+		select {
+		case <-time.Tick(interval):
```

Comment:
This will create and leak a new `time.Ticker` every iteration. See https://golang.org/pkg/time/#Tick
To avoid this, call `time.Tick(interval)` once outside the for loop and reuse the channel for each iteration.